runtime-sdk/src/modules/rofl/app/processor.rs

```diff
@@ -0,0 +1,186 @@
+use std::sync::Arc;
+
+use anyhow::{anyhow, Result};
+use tokio::sync::{mpsc, oneshot};
+
+use crate::{
+    core::{
+        common::logger::get_logger,
+        consensus::{roothash, verifier::Verifier},
+        dispatcher::PreInitState,
+        host::{self, Host as _},
+        identity::Identity,
+        protocol::Protocol,
+    },
+    crypto::signature::{secp256k1, Signer},
+};
+use rand::rngs::OsRng;
+
+use super::{App, Environment};
+
+/// Size of the processor command queue.
+const CMDQ_BACKLOG: usize = 32;
+
+/// Command sent to the processor task.
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub(super) enum Command {
+    /// Process a notification of a new runtime block.
+    ProcessRuntimeBlock(roothash::AnnotatedBlock),
+    /// Retrieve the latest known round.
+    GetLatestRound(oneshot::Sender<u64>),
+    /// Notification that initial registration has been completed.
+    InitialRegistrationCompleted,
+}
+
+/// Processor state.
+pub(super) struct State<A: App> {
+    pub(super) identity: Arc<Identity>,
+    pub(super) host: Arc<Protocol>,
+    pub(super) consensus_verifier: Arc<dyn Verifier>,
+    pub(super) signer: Arc<dyn Signer>,
+    pub(super) app: Arc<A>,
+}
+
+struct Subtasks {
+    registration: mpsc::Sender<()>,
+    notifier: mpsc::Sender<super::notifier::Notify>,
+}
+
+/// Processor task.
+pub(super) struct Task<A: App> {
```

Comment:
> This way you can construct the processor and tasks without needing to run them

The problem is that running the task requires it to be consumed (so it can be mutated independently). So in your case we would need to split each task into something like an _implementation_ and _handle_. This can be done but maybe it is a bit too much at this time, especially given that this is all internal and can change later without affecting the public API.

Reply:
Ok maybe it's not too bad, I'll go ahead and refactor this a bit.

Reply:
> The problem is that running the task requires it to be consumed (so it can be mutated independently). So in your case we would need to split each task into something like an implementation and handle.

In that case, maybe it is really the best to leave it as it is.

Reply:
Pushed a refactored version, please take another look.

Reply:
Looks nice. You were right, it is pita to implement `new`, `start`, `stop`.